### PR TITLE
Audit of auxinfo protocol

### LINKED ISSUE #275: Audit auxinfo protocol
Take a pass over the auxinfo module and identify concrete issues that should be addressed, according to the criteria in #272.

- [ ] Review auxinfo module
- [ ] Document findings and potential concerns
- [ ] Write specific issues and put them in #272

### PR DESCRIPTION
**Pull request by amaloz**
Closes #275.

This PR audits the auxinfo protocol, adding comments to clarify the protocol alongside making several small code improvements.

### CODE REVIEW

**Comment by marsella** [comment]
I'm seeing a [many-hour CI run](https://github.com/boltlabs-inc/tss-ecdsa/actions/runs/4702514098/jobs/8339868946?pr=298) with a hanging test:
> test auxinfo::participant::tests::test_run_auxinfo_protocol has been running for over 60 seconds

Cancelled the job for now.

**Comment by amaloz** [comment]
> I'm seeing a [many-hour CI run](https://github.com/boltlabs-inc/tss-ecdsa/actions/runs/4702514098/jobs/8339868946?pr=298) with a hanging test:
> 
> > test auxinfo::participant::tests::test_run_auxinfo_protocol has been running for over 60 seconds
> 
> Cancelled the job for now.

Oof, yeah I think it's the same issue (I was auditing this and `keygen` in parallel). Let me fix it.

File: src/auxinfo/participant.rs [hunk 1]
```diff
@@ -420,8 +485,9 @@ impl AuxInfoParticipant {
         let proof = AuxInfoProof::prove(
             rng,
             self.retrieve_context(),
-            message.id(),
+            sid,
             global_rid,
+            // XXX BUG! this should be the _other party_'s setup parameters!
             my_public.params(),
```
Comment (amaloz) [inline src/auxinfo/participant.rs:491]: I'm very confused why this passes verification: we pass our own setup parameters here, but as far as I can tell we use the other participant's setup parameters when verifying the proof below?

File: src/auxinfo/participant.rs [hunk 2]
```diff
@@ -83,6 +85,26 @@ pub enum Status {
...
+///   participants.
+/// - In the third round, each participant (1) checks the validity of all the
+///   commitments plus the validity of the committed `𝚷[prm]` proof, and (2)
+///   produces two additional proofs: the first, `𝚷[mod]`, which asserts the
+///   validity of `N`, and the second, `𝚷[fac]`, which ????. Each participant
+///   sends these proofs to all other participants.
+/// - Finally, in the last round each participant checks the validity of the
+///   proofs from round three. If everything passes, each participant outputs
+///   the `(N, s, t)` tuples from all other parties, alongside its own secret
+///   primes `(p, q)`.
```
Comment (indomitableSwan) [inline src/auxinfo/participant.rs:106]: Also its own `s, t`.

File: src/auxinfo/participant.rs (hunk 1 above)
Comment (indomitableSwan) [reply src/auxinfo/participant.rs:492]: This is really hard to follow, I agree. Maybe @tyurek is more familiar with what's going on here?

Comment (indomitableSwan) [reply src/auxinfo/participant.rs:492]: Oh, this does make sense. The sender creates `Pi_fac` using only their own AuxInfo parameters. The verifier than says "OK this proof should verify with the sender's AuxInfo parameters." It's wrong, but it's consistent.

Comment (amaloz) [reply src/auxinfo/participant.rs:492]: Ah yes! That does make sense.

Comment (marsella) [reply src/auxinfo/participant.rs:492]: I agree with swan. The self/other switch makes sense between prove and verify because different parties run those protocols, but there is still a bug because I think the prover should use the verifier's commitment parameters to create the proof.

File: src/auxinfo/participant.rs [hunk 3]
```diff
@@ -83,6 +85,27 @@ pub enum Status {
...
+///   alongside ring-Pedersen parameters `(s, t, λ)` such that `s = t^λ mod N`.
+///   Each participant then produces a zero-knowledge proof `𝚷[prm]` that the
+///   ring-Pedersen parameters are correct. Finally, each participant commits to
+///   the tuple `(N, s, t, 𝚷[prm])` and broadcasts this commitment.
+/// - Once all participants have received all broadcasted commitments, in the
+///   second round each participant sends its own decommitment to all other
+///   participants.
+/// - In the third round, each participant (1) checks the validity of all the
+///   commitments plus the validity of the committed `𝚷[prm]` proof, and (2)
+///   produces two additional proofs: the first, `𝚷[mod]`, which asserts the
```
Comment (indomitableSwan) [inline src/auxinfo/participant.rs:100]: We should fix this documentation, since it's not just a single `𝚷[fac]` proof, but instead one for each of the other participants.

Comment (amaloz) [reply src/auxinfo/participant.rs]: Good catch. Fixed, alongside a larger rewording to hopefully make this section clearer.

File: src/auxinfo/participant.rs (hunk 2 above)
Comment (amaloz) [reply src/auxinfo/participant.rs]: I added working to say the `(N, s, t)` tuple is from all parties (including ourselves), which hopefully addresses this point.

File: src/auxinfo/participant.rs [hunk 4]
```diff
@@ -522,6 +598,8 @@ impl AuxInfoParticipant {
         debug!("Creating new auxinfo.");
 
         let (decryption_key, p, q) = DecryptionKey::new(rng)?;
+        // XXX should this be `self.retrieve_context()` instead of `&()`? I
+        // realize they are the same now, but may change in the future?
```
Comment (marsella) [inline src/auxinfo/participant.rs:602]: Yes, it should be! Rawane fixed this in [an open PR](https://github.com/boltlabs-inc/tss-ecdsa/pull/280/files#diff-8464b562e103bfbf3b9654e47a7c24ee39f48127ee34eca29318e310527f505cR527) but you can also fix it here to avoid a merge conflict mistake if you want.

File: src/auxinfo/proof.rs [hunk 5]
```diff
@@ -38,34 +41,35 @@ impl AuxInfoProof {
...
             &PiModSecret::new(p, q),
             context,
             &mut pimod_transcript,
             rng,
         )?;
-        let mut pifac_transcript = Transcript::new(b"PiFacProof");
-        pifac_transcript.append_message(b"PiFac ProofContext", context.as_bytes());
-        pifac_transcript.append_message(b"Session Id", &serialize!(&sid)?);
-        pifac_transcript.append_message(b"rho", &rho);
+        let mut pifac_transcript = Self::pifac_transcript(context, sid, rho)?;
```
Comment (marsella) [inline src/auxinfo/proof.rs:70]: I think this should be appended on top of the transcript for pimod, actually. You can do that here or make a separate issue (I'm sensing an "improve code quality in AuxInfoProof" ticket)

**Review by marsella — CHANGES_REQUESTED** [review]
Great audit, thanks! I have some questions and comments about the issues you identified as well as some code changes requested on the fixes you made.

File: src/auxinfo/auxinfo_commit.rs [hunk 6]
```diff
@@ -33,14 +36,19 @@ impl AuxInfoCommit {
     }
 }
 
+/// Commitment scheme for the auxinfo protocol.
+///
+/// XXX: Calling this `...Decommit` is confusing I think, since this both
+/// produces commitments and decommits them. It should really be called
+/// `(AuxInfo)CommitmentScheme` I think.
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/auxinfo_commit.rs:43]: I'll ad a ticket for this and for now re-name `AuxInfoDecommit` to `CommitmentScheme`

File: src/auxinfo/participant.rs [hunk 7]
```diff
@@ -144,20 +169,19 @@ impl ProtocolParticipant for AuxInfoParticipant {
...
-            MessageType::Auxinfo(AuxinfoMessageType::Ready) => self.handle_ready_msg(rng, message),
             MessageType::Auxinfo(AuxinfoMessageType::R2Decommit) => {
-                self.handle_round_two_msg(rng, message, input)
+                self.handle_round_two_msg(rng, message)
             }
             MessageType::Auxinfo(AuxinfoMessageType::R3Proof) => {
-                self.handle_round_three_msg(rng, message, input)
+                self.handle_round_three_msg(message)
             }
-            MessageType::Auxinfo(_) => Err(InternalError::MessageMustBeBroadcasted),
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:168]: Yep, looks like you beat me to it!

File: src/auxinfo/participant.rs [hunk 8]
```diff
@@ -293,63 +336,77 @@ impl AuxInfoParticipant {
...
+            // produce the necessary public info we were looking for above.
+            let more_messages = run_only_once!(self.gen_round_one_msgs(rng, sid))?;
             messages.extend_from_slice(&more_messages);
         }
 
-        // retrieve your decom from storage
         let decom = self.local_storage.retrieve::<storage::Decommit>(self.id)?;
         let decom_bytes = serialize!(&decom)?;
-        let more_messages: Vec<Message> = self
+        let messages: Vec<Message> = self
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:365]: Good catch!

File: src/auxinfo/participant.rs [hunk 9]
```diff
@@ -293,63 +336,77 @@ impl AuxInfoParticipant {
...
         if !r1_done {
-            // store any early round2 messages
+            // store any early round 2 messages
             self.stash_message(message)?;
             return Ok(ProcessOutcome::Incomplete);
         }
+        // Convert the message into the decommitment value.
+        //
+        // Note: `AuxInfoDecommit::from_message` checks the validity of all its
+        // messages, which includes validating the `𝚷[prm]` proof.
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:409]: Agreed, it was quite confusing when auditing that the `from_message` functions also did validation. I'll add an issue!

[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:409]: See here: https://github.com/boltlabs-inc/tss-ecdsa/issues/347

File: src/auxinfo/participant.rs [hunk 10]
```diff
@@ -358,19 +415,24 @@ impl AuxInfoParticipant {
...
+        // could be in the setting where we've received round two messages from
+        // all other participants but haven't yet generated our own round one
+        // message.
         let r2_done = self
             .local_storage
             .contains_for_all_ids::<storage::Decommit>(&self.other_participant_ids);
         if r2_done {
             // Generate messages for round 3...
-            let round_two_messages = run_only_once!(self.gen_round_three_msgs(rng, message))?;
+            let round_two_messages = run_only_once!(self.gen_round_three_msgs(rng, message.id()))?;
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:429]: I'm just surprised I missed all of these!

File: src/auxinfo/proof.rs (hunk 5 above)
Comment (amaloz) [reply src/auxinfo/proof.rs:76]: Not sure I follow. Are you saying we should have a single transcript for both proofs?

File: src/auxinfo/participant.rs [hunk 11]
```diff
@@ -416,44 +482,50 @@ impl AuxInfoParticipant {
...
+                    rng,
+                    self.retrieve_context(),
+                    sid,
+                    global_rid,
+                    // XXX: Replace `into_public` with reference to
+                    // `AuxInfoPublic`? That should avoid an unnecessary clone.
+                    verifier_decommit.clone().into_public().params(),
+                    &product,
+                    &witness.p,
+                    &witness.q,
```
[the remark this answers is not in the document]
Comment (amaloz) [reply src/auxinfo/participant.rs:539]: Added a comment to #53 to log this point. Thanks!

File: src/auxinfo/proof.rs (hunk 5 above)
Comment (marsella) [reply src/auxinfo/proof.rs:76]: Yes.

Comment (marsella) [reply src/auxinfo/proof.rs:76]: E.g. the new `pifac_transcript` method should take a transcript as a parameter and update it, rather than making a new one from scratch.

**Review by marsella — COMMENTED** [review]
Thanks, this looks great! I reviewed the issues as well and they also look good.

The only remaining item for me is fixing the transcripts per the open conversation.

File: src/auxinfo/proof.rs (hunk 5 above)
Comment (amaloz) [reply src/auxinfo/proof.rs]: Done (I think, please take a look :-) )

Comment (marsella) [reply src/auxinfo/proof.rs]: Yes, perfect. Thank you!
